Actualize the API documentation style guide

patiencedaur opened the pull request:
Deployment: https://docs.d.tarantool.io/en/doc/api-style-guide/contributing/docs/api/

patiencedaur committed:
Provide a better title for API documentation guidelines

patiencedaur committed:
Link to the 'Language and style' section

patiencedaur committed:
Elaborate on best style practices

Totktonada reviewed (approved):
LGTM. I would also add `CONTRIBUTING.md` file, because GitHub highlight it for first-time contributors. It is good entry point for new contributors.

p7nov commented on doc/contributing/docs/api.rst:
```diff
@@ -82,7 +88,7 @@ Configuration parameters
 
 *   Since which Tarantool version
-*   General description
+*   General description (in imperative mood)
```
Minor: "imperative mood" seems unclear for parameter descriptions – they are mostly noun phrases.

veod32 replied:
It depends.
In many cases, we describe parameters starting with the verb and using the imperative mood. For example,

- https://www.tarantool.io/en/doc/latest/reference/configuration/#confval-background
- https://www.tarantool.io/en/doc/latest/reference/configuration/#confval-background

But in some cases the description starts with a noun, like
https://www.tarantool.io/en/doc/latest/reference/configuration/#confval-memtx_dir
https://www.tarantool.io/en/doc/latest/reference/configuration/#confval-memtx_dir

Perhaps, we need to recommend both options to writes and give examples of both.

veod32 commented on doc/contributing/docs/api.rst:
```diff
@@ -82,7 +88,7 @@ Configuration parameters
 For every configuration parameter, list the following details:
 
 *   Since which Tarantool version
```
We need to change the guideline and example by explaining how to write the version info -- use Sphinx role or reference to the corresponding release note (RN) document, for example,
```
Since :doc:`2.10.0 </release/2.10.0>`.
```

patiencedaur committed:
Apply review suggestions and restructure the guide

patiencedaur committed:
Add a CONTRIBUTING file

patiencedaur commented:
We might also consider explaining how the `..  function::` and `..  module::` directives work, specifically that they don't produce headings displayed in the right menu. This means that each new function section should start with an `.. _anchor:`, which should:
- either be linked to in the "List of all functions"/"Index" section on the module page
- or put in the hidden `.. toctree::` on the module page.

Let's do it in another PR.

veod32 commented on doc/contributing/docs/api.rst:
```diff
@@ -0,0 +1,170 @@
+Indicating the version
+----------------------
+
+For every new feature or parameter, please indicate the version it was introduced in.
```
(here and below in the text)

"Feature" sounds a bit unclear: actually, it can be anything -- a new module, or function, or method, or even a new parameter. Maybe we can mention them more specific?

Besides, I think we should be more specific when we need to mention the version and when we don't. 

For example, a new module is implemented in the 2.10.0 release. In the module description, we definetely should mention "Since 2.10.0". Should we do the same for each of the module functions and methods? Seemingly yes, because a user can navigate directly to a function/method page without looking at the module description. 
Should we mention "Since 2.10.0" for each parameter of the new module? I think no, because it's excessive -- this info is already here in the function/module description. We need to specify a version for a parameter if this parameter is "a feature" and the version it's introduced in differs from the version of the function/method and all other parameters.

veod32 commented on doc/contributing/docs/api.rst:
```diff
@@ -0,0 +1,170 @@
+    Since :tarantool-release:`2.10.0`.
+    This is a link to the release notes on GitHub.
+
+    Since :doc:`2.10.0 </release/2.10.0>`.
```
I think we need to use only this one in API description. Otherwise, we can get a kind of mess: someone will put a GitHub link and someone -- a link to documentation RNs. Better to unify this:
- in API section of documentation, we give links to RNs in our documentation
- in RNs, we give links to the original GutHub release notes (:tarantool-release:)

veod32 commented on doc/contributing/docs/api.rst:
```diff
@@ -0,0 +1,170 @@
+*   Start with a verb in the imperative mood. Example: *Create a fiber.*
+*   Start with a noun. Example: *The directory where memtx stores snapshot files.*
+
+Checklist
```
I would add a short intro: that each list item is a point that should be described (taking into account that some items can be optional, of course).

veod32 commented on doc/contributing/docs/api.rst:
```diff
@@ -0,0 +1,170 @@
+*   :ref:`Parameters <documenting_parameters>`
+*   What this function returns (if nothing, write 'none')
+*   Return type (if exists)
+*   Possible errors (if exist)
```
I would make a separate section for "Possible errors" in our guidelines similar to the "Function and class parameters" section.

patiencedaur replied:
There's a separate issue for this: https://github.com/tarantool/doc/issues/1810

veod32 replied:
ОК

veod32 commented on doc/contributing/docs/api.rst:
```diff
@@ -0,0 +1,170 @@
+*   What this function returns (if nothing, write 'none')
+*   Return type (if exists)
+*   Possible errors (if exist)
+*   Complexity factors (if exist)
```
We can make a separate section in the guidelines or mention directly here that complexity factors are relevant for a few functions only, and put a link here to https://www.tarantool.io/en/doc/latest/book/box/data_model/#complexity-factors

veod32 commented on doc/contributing/docs/api.rst:
```diff
@@ -0,0 +1,170 @@
+
+..  _documenting_parameters:
+
+Function and class parameters
```
Maybe "Function and method parameters"? I don't remember if a class itself can have parameters. [Not sure about that]

patiencedaur replied:
Oops, that was definitely a typo.

veod32 commented on doc/contributing/docs/api.rst:
```diff
@@ -0,0 +1,170 @@
+*   Type
+*   Default value (if optional), possible values
+
+If the parameter is optional, make sure it is enclosed in square brackets
+in the function declaration (in the "heading").
```
It's good that this guideline is in general description for all the parameters now!

Maybe we can give an example here? Because it's not clear what "heading" means in this case.

Also, we can think of being more direct: to tell the guys not to write "Optional" and "Required" in the paramter description because you can see that in the function declaration.

veod32 commented on doc/contributing/docs/api.rst:
```diff
@@ -0,0 +1,170 @@
+in the function declaration (in the "heading").
+
+In the "Possible errors" section of the function or class method,
+consider explaining what happens if the parameter hasn't been defined or has the wrong value.
```
Can we put this in a separate section in the guidelines? Because possible errors can be related to the entire function/method, not necesserily to the misuse of some parameters. 
See, for example, 
- https://www.tarantool.io/en/doc/latest/book/connectors/cxx/tntcxx_api/#public-methods
- https://www.tarantool.io/en/doc/latest/reference/reference_lua/box_space/pairs/

veod32 commented on doc/contributing/docs/api.rst:
```diff
@@ -0,0 +1,170 @@
+*   Type
+*   Corresponding environment variable (if applicable)
+*   Default value
+*   Possible values
```
In the "Examples and templates > Configuration parameters" below, let's make it more clear where to describe possible values: in the general description or as a separate item like 
Type
Env variable
Default value
Dynamic

veod32 commented on doc/contributing/docs/api.rst:
```diff
@@ -0,0 +1,170 @@
+We use the Sphinx directives ``.. module::``
+and ``.. function::`` to describe functions of Tarantool modules:
+
+..  literalinclude:: ./_includes/function_template.rst
```
In this example we have
```
    :param function-arguments: what will be passed to function.
                               If the arguments are optional, put them in
                               square brackets in the function declaration.
```

I suggest to remove "If the arguments are optional, put them in square brackets in the function declaration."  
- We have this guideline already in the "Function and class parameters" section.
- It's now a mixture of the actual functional description of the parameter and the meta info (writing guidelines). And it's confusing.

veod32 commented on doc/contributing/docs/api.rst:
```diff
@@ -0,0 +1,170 @@
+
+Example:
+
+..  literalinclude:: ./_includes/confval_template.rst
```
Maybe we can give here another example? Of a parameter with a number of possible values -- to illustrate this checklist item.

patiencedaur replied:
https://www.tarantool.io/en/doc/latest/reference/configuration/#confval-election_mode
This is the only one with "possible values" provided in a list. It has a huge description, so I wouldn't include it in this specific article.
I think that the possible values may, but need not, be a list.

patiencedaur committed:
Apply suggestions from review

* Separate version formatting guidelines for RN and API

  Use :tarantool-release: in release notes
  and :doc:`x.y.z </release/x.y.z>` in API reference.

* Add complexity factor note and links

* Fix typos and inconsistencies

* Give an example of optional parameters in square brackets

* Make a stub "Possible errors" section

* Remove redundant information from function example
